### PR TITLE
Fix for #5808: Update the cached Place Items on the successful association of Wiki Item

### LINKED ISSUE #5808: Nearby: invalidate cache after uploading from Nearby
The pin does not turn from red to green after upload.

Steps to reproduce:
1. Open Nearby
2. Select a red pin
3. Upload a picture for it
4. Close and reopen the app, open Nearby again
5. Problem: The pin is still red. It looks like I need to upload again.

Fix: The app should remove that pin's cache item after the upload.

If easier, removing all pins from the cache is OK.

### PR DESCRIPTION
**Pull request by ChristoJobyAntony**
When a user uploads a valid depiction of a Wiki Item which currently has no image (represented by a red pin), the pin should turn green on the successful upload and depiction of the image. However, due to caching of Nearby places, the pin colour is not updated after user's image upload.

Fixes #5808

### Changes
1. ~~A new method was added to the `PlacesRepository` to delete a given `Place` entity.~~
2. ~~On the successful claiming of a `WikiItemEdit`, the corresponding Wiki Item is cleared from the cache~~
3. On the successful claiming of a `WikiItemEdit`, the corresponding Wiki Item updated in the local cache.
4. The updated link is created using the format: `HOME_URL + uploadResult.createCanonicalFileName()`
5. The field `name` of this `Place` object is also updated from the `WikiItem`. This is because there was a scenario where a `Place` fetched would have an empty name. Therefore, the `NearbyParentFragment` tries to update Places without a valid `name` from the web, leading to a race condition. This situation was observed through logs. 
image

Tested  `ProdDebug` on Pixel 7 with API level 34.
https://github.com/user-attachments/assets/6328b8ca-8342-453b-815a-30f0c10d36f2

### CODE REVIEW

File: app/src/main/java/fr/free/nrw/commons/upload/worker/UploadWorker.kt [hunk 1]
```diff
@@ -379,7 +383,7 @@ class UploadWorker(
                                 saveCompletedContribution(contribution, uploadResult)
                             } else {
                                 Timber.d(
-                                    "WikiDataEdit not required, making wikidata edit",
+                                    "WikiDataEdit required, making wikidata edit",
```
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/upload/worker/UploadWorker.kt:386]: Thanks!

File: app/src/main/java/fr/free/nrw/commons/nearby/PlacesLocalDataSource.java [hunk 2]
```diff
@@ -36,4 +36,11 @@ public Place fetchPlace(String entityID){
     public Completable savePlace(Place place) {
         return placeDao.save(place);
     }
+
+    /**
+     * Deletes a Place object from the database
+     * based on the provided entity.
+     */
+    public void  deletePlace(final Place place) {placeDao.deletePlace(place);}
```
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/nearby/PlacesLocalDataSource.java:44]: nitpick: double space, please make it single space

**Review by nicolas-raoul — CHANGES_REQUESTED** [review]
Is it still a draft?

Fantastic code, and screencast looks great!

File: app/src/main/java/fr/free/nrw/commons/nearby/PlacesLocalDataSource.java (hunk 2 above)
Comment (ChristoJobyAntony) [reply app/src/main/java/fr/free/nrw/commons/nearby/PlacesLocalDataSource.java:44]: Sure will do ! would you want me to add this as a new commit or revert and edit the previous commits ?

Comment (nicolas-raoul) [reply app/src/main/java/fr/free/nrw/commons/nearby/PlacesLocalDataSource.java:44]: New commit is fine, before merging I will squash all of this branch's commits into a single commit anyway. :-)

**Comment by ChristoJobyAntony** [comment]
> Is it still a draft?

@nicolas-raoul, I had left the PR as draft to get your feedback on the approach I had taken. If you think this is approach is acceptable, I am happy to set this PR to ready to review.

File: app/src/main/java/fr/free/nrw/commons/nearby/PlacesLocalDataSource.java (hunk 2 above)
Comment (ChristoJobyAntony) [reply app/src/main/java/fr/free/nrw/commons/nearby/PlacesLocalDataSource.java]: Thank you, I have pushed the changes.

File: app/src/main/java/fr/free/nrw/commons/upload/worker/UploadWorker.kt (hunk 1 above)
Comment (ChristoJobyAntony) [reply app/src/main/java/fr/free/nrw/commons/upload/worker/UploadWorker.kt:386]: Haha, I was really confused by this log before the Edit was made. Realised it was a typo from the previous log.

**Review by nicolas-raoul — COMMENTED** [review]
Code looks great!
I am testing on my device, now going outside to actually get pictures of places and try it out in the real world. :-)

**Comment by nicolas-raoul** [comment]
Somehow the pin stayed red in my test (screencast below)... any idea what might be happening?

Maybe a server-side cache issue? Analyzing server response to our query_for_item.rq request would tell. In that case deleting the cache again every 10 seconds for 1 minute may be a workaround.

https://github.com/user-attachments/assets/65f45789-08d1-4173-8211-021fd8140cc1

**Comment by ChristoJobyAntony** [comment]
> Somehow the pin stayed red in my test (screencast below)... any idea what might be happening?

Oh, that is very interesting. The only difference I can possibly adhere to is that I had added a breakpoint at the step to ensure the cache was being cleared. This could have possibly added enough delay for the server side response to be updated. 

>Maybe a server-side cache issue? Analyzing server response to our query_for_item.rq request would tell. In that case deleting the cache again every 10 seconds for 1 minute may be a workaround.

Alternatively, do you think it would be appropriate to instead directly write the updated image link to the cache, this will only be done if a `revisionID` is returned. It would reduce the requests required to be sent out.

**Comment by nicolas-raoul** [comment]
> Alternatively, do you think it would be appropriate to instead directly write [...] to the cache

Usually it is more elegant (even if a bit wasteful) but only clear the cache and then let the normal code load the new value from the server. But in this case I agree it is a good way to avoid server-side cache issues.

Looking forward to the updated pull request, thanks a lot for the great idea!

**Comment by ChristoJobyAntony** [comment]
@nicolas-raoul could you revert the merge, I had not pushed the updated name code into this branch yet.

**Comment by ChristoJobyAntony** [comment]
@nicolas-raoul I have updated the PR description, and also updated the code to reflect our conversation.  Would you be able to review this PR ?

**Review by nicolas-raoul — APPROVED** [review]
Works great, thanks a lot!

https://github.com/user-attachments/assets/f8c29bb2-4be8-4b08-8c9f-b20bfabcc723

**Pull request merged**